sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -313,6 +314,42 @@ trait CheckAnalysis extends PredicateHelper {
               failAnalysis(s"Invalid partitioning: ${badReferences.mkString(", ")}")
             }
 
+          case alter: AlterTable if alter.childrenResolved =>
+            val table = alter.table
+            def findField(operation: String, fieldName: Array[String]): StructField = {
+              // include collections because structs nested in maps and arrays may be altered
+              val field = table.schema.findNestedField(fieldName, includeCollections = true)
+              if (field.isEmpty) {
+                throw new AnalysisException(
+                  s"Cannot $operation missing field in ${table.name} schema: ${fieldName.quoted}")
+              }
+              field.get
+            }
+
+            alter.changes.foreach {
+              case add: AddColumn =>
+                val parent = add.fieldNames.init
+                if (parent.nonEmpty) {
+                  findField("add to", parent)
+                }
+              case update: UpdateColumnType =>
+                val field = findField("update", update.fieldNames)
+                if (!Cast.canUpCast(field.dataType, update.newDataType)) {
```

Comment:
It would be good to have test cases for the upcast failure

Reply:
Also, it seems that update nested column entirely should fail, right? E.g.
```
CREATE TABLE t (id int, points map<struct<x: float, y: double>, bigint>) USING foo
ALTER TABLE t ALTER COLUMN points TYPE map<struct<x: float, y: double, z: double>, bigint>
```

Reply:
There is a test for upcast failure. See "AlterTable: update column type must be compatible".

Why should altering a nested column fail? As long as the new column is optional, there is no problem.

Reply:
Okay, I agree that you should not be able to alter complex types with another complex type

Reply:
@gengliangwang, I've updated this and added tests for these cases. Thanks for catching the complex type case.